### PR TITLE
Workaround on cusparse SPMM broadcasting issue in cuda 11.6 update 1

### PR DESCRIPTION
**Pull request by xwang233**
Starting cuda 11.6 update 1, cusparse CSR SPMM may silently give wrong output when the input sparse CSR matrix A is broadcasted, by setting both strides to 0 in `cusparseCsrSetStridedBatch` call.

While the issue has been fixed internally, it's necessary to have a workaround in pytorch repo to fix this issue before the public cuda release. This PR "fixes" the issue in pytorch by doing batched CSR SPMM on each matrix separately in a `for` loop.

See also https://github.com/pytorch/pytorch/pull/68711

### CODE REVIEW

**Review by IvanYashchuk — CHANGES_REQUESTED** [review]
I'm trying to understand what's the exact problem introduced in CUDA 11.6 Update 1.

Is that right that the problem occurs only for the case when A is a single matrix and B is batched, i.e. the first batch mode from the documentation:
image

Also, I'd like to note that the other two batch modes are known not to work despite the documentation, and it's only mentioned in a comment in the code example from NVIDIA:
https://github.com/NVIDIA/CUDALibrarySamples/blob/1c825f412455a52f718aa9425e3b5574d17efff5/cuSPARSE/spmm_csr_batched/spmm_csr_batched_example.c#L160
It could be fixed in recent versions but didn't try.

Does the broadcasting of matrix A work in CUDA Toolkit 11.6 Update 2?

If it's known to be broken only for CUDA Toolkit 11.6 Update 1, instead of workarounds that leads to less comprehensive code, I'd suggest simply checking for CUDA version and raising an error.

**Comment by IvanYashchuk** [comment]
This problem is not in the "Known issues" part of the release notes, is it?

**Comment by IvanYashchuk** [comment]
I compiled PyTorch with CUDA 11.6 Update 1 and I see failures for complex numbers but everything else passes:
```
test/test_sparse_csr.py::TestSparseCSRCUDA::test_baddbmm_cuda_complex128 FAILED                                                                                                                                                                                                  [ 56%]
test/test_sparse_csr.py::TestSparseCSRCUDA::test_baddbmm_cuda_complex64 FAILED                                                                                                                                                                                                   [ 62%]
test/test_sparse_csr.py::TestSparseCSRCUDA::test_baddbmm_cuda_float32 PASSED                                                                                                                                                                                                     [ 68%]
test/test_sparse_csr.py::TestSparseCSRCUDA::test_baddbmm_cuda_float64 PASSED                                                                                                                                                                                                     [ 75%]
test/test_sparse_csr.py::TestSparseCSRCUDA::test_bmm_cuda_complex128 PASSED                                                                                                                                                                                                      [ 81%]
test/test_sparse_csr.py::TestSparseCSRCUDA::test_bmm_cuda_complex64 PASSED                                                                                                                                                                                                       [ 87%]
test/test_sparse_csr.py::TestSparseCSRCUDA::test_bmm_cuda_float32 PASSED                                                                                                                                                                                                         [ 93%]
test/test_sparse_csr.py::TestSparseCSRCUDA::test_bmm_cuda_float64 PASSED
```

**Comment by xwang233** [comment]
The functionality failure occurs when sparse matrix A is broadcasted. Yes, you're right. We've only seen failures on complex inputs.

It's not in the "known issue" section.

Unfortunately, this issue started from cuda 11.6 update 1 and will still be existed in cuda 11.7 release. The fix will be targeted in the next minor release after 11.7 (cuda 11.7 update 1, if any).

I'm ok to fix this either way in pytorch: error out in those affected versions, or add a loop workaround. Also wanted to hear what @ngimel thinks, and maybe @cpuhrsch @ptrblck .

**Comment by xwang233** [comment]
ping @ngimel and @IvanYashchuk for the review

File: aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp [hunk 1]
```diff
@@ -651,6 +651,46 @@ void spmm(
...
+    const Tensor& result) {
+
+// Cusparse CSR SPMM, starting from cuda 11.6 update 1 (cusparse 11.7.2), has some functionality issue when sparse matrix A is being broadcasted,
+//   by setting both strides to 0 in cusparseCsrSetStridedBatch call.
+// This will be fixed in a future cuda release.
+// We are using a temporary workaround here to do the batched SPMM on every single matrix separately.
+#if CUSPARSE_VERSION < 11702 || CUSPARSE_VERSION >= 11704
+  return spmm(mat1, mat2, beta, alpha, result);
+#else
+  if (mat1.dim() == 2) {
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:668]: As we found out the bug is present only for complex inputs. Let's use the workaround only for complex inputs then. I expect a for-loop solution to be slower and no need to slow down the code path for reals if it works fine. A comment above should include info that the bug is present only for complex inputs.
```suggestion
  if (mat1.dim() == 2 || mat1.is_floating_point()) {
```

**Review by IvanYashchuk — CHANGES_REQUESTED** [review]
Thank you for identifying the problem and proposing the workaround! I think it's better to have it only for complex inputs.

File: aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp [hunk 2]
```diff
@@ -651,6 +651,46 @@ void spmm(
...
+        mat1.col_indices().dim() >= 2) {
+      // cuSPARSE ignores the strides and uses only the first batch, see #68711
+      TORCH_INTERNAL_ASSERT(
+          false,
+          "Support for batched CSR indices and values is not implemented.");
+    }
+
+    auto mat1_copy = mat1.clone();
+    static_cast<SparseCsrTensorImpl*>(mat1_copy.unsafeGetTensorImpl())->
+      set_sizes_contiguous(mat1_copy.sizes().slice(sparse_dim - 2));
```
[the remark this answers is not in the document]
Comment (xwang233) [reply aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:684]: Thanks! Looking forward to the update on CSR descriptor!

File: aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp [hunk 3]
```diff
@@ -651,6 +651,46 @@ void spmm(
...
+
+    int batch_count = at::native::cuda_int_cast(at::native::batchCount(mat1), "batch_count");
```
[the remark this answers is not in the document]
Comment (xwang233) [reply aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:686]: I was following the existing code in `CuSparseSpMatCsrDescriptor` constructor to make the code base look consistent.

**Pull request closed without merging**